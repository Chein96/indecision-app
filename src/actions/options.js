import database from '../firebase/firebase';

export const addOption = (option) => ({
    type: 'ADD_OPTION',
    option
});

export const startAddOption = (optionData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const { description = '' } = optionData;
        const option = { description };
        
        return database.ref(`users/${uid}/options`).push(option).then((ref) => {
            dispatch(addOption({
                id: ref.key,
                ...option
            }));
        });
    }
};

export const removeOption = ({ id }) => ({
    type: 'REMOVE_OPTION',
    id
});

export const startRemoveOption = ({ id }) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/options/${id}`).remove().then(() => {
            dispatch(removeOption({ id }));
        });
    }
};

export const removeAllOptions = () => ({
    type: 'REMOVE_ALL_OPTIONS'
});

export const startRemoveAllOptions = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/options`).set([]).then(() => {
            dispatch(removeAllOptions());
        });
    }
};

export const setOptions = (options) => ({
    type: 'SET_OPTIONS',
    options
});

export const startSetOptions = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/options`).once('value').then((snapshot) => {
            const optionsData = [];
            snapshot.forEach((childSnapshot) => {
                optionsData.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setOptions(optionsData));
        });
    }
}