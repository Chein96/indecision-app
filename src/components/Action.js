import React, { Fragment } from 'react';
import { connect } from 'react-redux';

export const Action = ({ handlePick, hasOptions }) => (
    <Fragment>
        <button
            className="big-button"
            disabled={!hasOptions}
            onClick={handlePick}
        >
            What should I do?
        </button>
    </Fragment>
);

const mapStateToProps = ({ options }) => ({
    hasOptions: options.length > 0
});

export default connect(mapStateToProps)(Action);