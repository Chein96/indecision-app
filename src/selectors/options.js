const optionExists = (options, option) => {
    const result = options.filter(({ description }) => {
        if(description === option){
            return true;
        }
    });
    if(result.length > 0){
        return true;
    }
    else
    {
        return false;
    }
};

export default optionExists;