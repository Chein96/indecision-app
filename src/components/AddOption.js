import React, { Fragment, useState } from 'react';
import { connect } from 'react-redux';
import { startAddOption } from '../actions/options';
import optionExists from '../selectors/options';

export const AddOption = ({ options, startAddOption }) => {
    const [ error, setError ] = useState(undefined);

    const handleAddOption = (e) => {
        e.preventDefault();
        const description = e.target.elements.option.value.trim();

        if(description === ''){
            setError('Enter a valid value to add item');
        }
        else if(optionExists(options, description)){
            setError('This option already exists');
        }
        else {
            const option = { description };
            startAddOption(option);
            e.target.elements.option.value = '';
            setError(undefined);
        }
    }

    return (
        <Fragment>
            {
                error
                &&
                <p className="add-option-error">{error}</p>
            }
            <form className="add-option" onSubmit={handleAddOption}>
                <input className="add-option__input" type="text" name="option"/>
                <button className="button">Add Option</button>
            </form>
        </Fragment>
    )
}

const mapStateToProps = ({ options }) => ({
    options
});

const mapDispatchToProps = (dispatch) => ({
    startAddOption: (option) => dispatch(startAddOption(option))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddOption);