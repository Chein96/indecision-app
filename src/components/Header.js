import React from 'react';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({startLogout}) => (
    <div className="header">
        <div className="container header__content">
            <div>
                <h1 className="header__title">Indecision</h1>
                <h2 className="header__subtitle">
                    Put your life in the hands of a computer
                </h2>
            </div>
            <button className="button button--link" onClick={startLogout}>Logout</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);