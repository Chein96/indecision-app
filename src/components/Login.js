import React from 'react';
import { connect } from 'react-redux';
import { startGoogleLogin } from '../actions/auth';

export const Login = ({ startGoogleLogin }) => (
    <main className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout-title">Indecision</h1>
            <p>Put your life in the hands of a computer.</p>
            <button className="button" onClick={startGoogleLogin}>Login with Google</button>
        </div>
    </main>
);

const mapDispatchToProps = (dispatch) => ({
    startGoogleLogin: () => dispatch(startGoogleLogin())
})

export default connect(undefined, mapDispatchToProps)(Login);