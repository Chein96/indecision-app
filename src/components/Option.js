import React from 'react';
import { connect } from 'react-redux';
import { startRemoveOption } from '../actions/options';

export const Option = ({ option, count, startRemoveOption }) => (
    <div className="option">
        <p className="option__text">{count}. {option.description}</p>
        <button
            className="button button--link"
            onClick={
                () => startRemoveOption({ id: option.id })
            }
        >
            Remove
        </button>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startRemoveOption: ({ id }) => dispatch(startRemoveOption({ id }))
});

export default connect(undefined, mapDispatchToProps)(Option);