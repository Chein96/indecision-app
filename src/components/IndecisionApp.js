import React from 'react';
import { connect } from 'react-redux';
import Action from './Action';
import AddOption from './AddOption';
import Header from './Header';
import Login from './Login';
import OptionModal from './OptionModal';
import Options from './Options';

class IndecisionApp extends React.Component {
    state = {
        selectedOption: undefined
    };

    handlePick = () => {
        const options = this.props.options;
        const randomNum = Math.floor(Math.random()*options.length);
        const option = options[randomNum];

        this.setState(() => ({
            selectedOption: option.description
        }));
    }

    handleClearSelectedOption = () => {
        this.setState(() => ({selectedOption: undefined}));
    }

    render() {
        return (
            this.props.uid ?
            (
                <div>
                    <Header />
                    <div className="container">
                        <Action
                            handlePick={this.handlePick}
                        />
                        <div className="widget">
                            <Options />
                            <AddOption />
                        </div>
                    </div>
                    <OptionModal
                        selectedOption={this.state.selectedOption}
                        handleClearSelectedOption={this.handleClearSelectedOption}
                    />
                </div>
            )
            :
            (
                <Login />
            )
        )
    }
};

const mapStateToProps = ({ options, auth }) => ({
    options,
    uid: auth.uid
});

export default connect(mapStateToProps)(IndecisionApp);