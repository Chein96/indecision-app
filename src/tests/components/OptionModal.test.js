import React from 'react';
import { shallow } from 'enzyme';
import Modal from 'react-modal';
import { OptionModal } from '../../components/OptionModal';
import options from '../fixtures/options';

let wrapper, selectedOption, handleClearSelectedOption;
beforeEach(() => {
    selectedOption = options[0].description;
    handleClearSelectedOption = jest.fn();
    wrapper = shallow(<OptionModal selectedOption={selectedOption} handleClearSelectedOption={handleClearSelectedOption} />);
});

test('Should render OptionModal correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle handleClearSelectedOption on button click', () => {
    wrapper.find('button').simulate('click');
    expect(handleClearSelectedOption).toHaveBeenCalled();
});

test('Should handle handleClearSelectedOption on user request', () => {
    wrapper.find(Modal).prop('onRequestClose')();
    expect(handleClearSelectedOption).toHaveBeenCalled();
});