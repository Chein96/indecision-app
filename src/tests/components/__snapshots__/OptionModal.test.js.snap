// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render OptionModal correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Selected Option"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Selected Option
  </h3>
  <p
    className="modal__body"
  >
    Buy milk
  </p>
  <button
    className="button"
    onClick={[Function]}
  >
    Okay
  </button>
</Modal>
`;
